devito/types/dense.py

```diff
@@ -84,6 +85,12 @@ def __init_finalize__(self, *args, function=None, **kwargs):
 
         # Data initialization
         initializer = kwargs.get('initializer')
+
+        # Don't want to reinitialise array if DataReference used as allocator;
+        # create a no-op intialiser to avoid overwriting the original array.
+        if isinstance(self._allocator, DataReference):
```

Comment:
can this be made part of the if-elif-else cascade below?

Reply:
No, several branches in the cascade use `initialiser` which means it needs to be set here to avoid code duplication. Keeping it here also results in more readable code.